Modules/_sre.c

```diff
@@ -248,22 +248,38 @@ data_stack_grow(SRE_STATE* state, Py_ssize_t size)
 /* -------------------------------------------------------------------- */
 /* factories and destructors */
 
-/* see sre.h for object declarations */
-static PyObject*pattern_new_match(PatternObject*, SRE_STATE*, Py_ssize_t);
-static PyObject *pattern_scanner(PatternObject *, PyObject *, Py_ssize_t, Py_ssize_t);
+/* module state */
+typedef struct {
+    PyTypeObject *Pattern_Type;
+    PyTypeObject *Match_Type;
+    PyTypeObject *Scanner_Type;
+} sre_module_state;
+
+static sre_module_state *
+get_sre_module_state(PyObject *m)
+{
+    sre_module_state *state = (sre_module_state *)PyModule_GetState(m);
+    assert(state);
+    return state;
+}
 
+static struct PyModuleDef sremodule;
+#define get_sre_module_state_by_type(tp) \
+    (get_sre_module_state(_PyType_GetModuleByDef(tp, &sremodule)))
```

Comment:
None of the types in the module have ``Py_TPFLAGS_BASETYPE``. AFAIK only subclass-able types need ``_PyType_GetModuleByDef``.

Reply:
This is used to fetch the state from slot methods. Is there a better way?

Reply:
I'm convinced that it's safe to use ``get_sre_module_state_by_class`` everywhere in this module. The left hand argument of ``richcompare`` is always an instance of the type. Python does not let users pass another type. The argument clinic lines are safe to use ``PyType_GetModule()``, too.

``_PyType_GetModuleByDef()`` is only required when a C extension type has the ``Py_TPFLAGS_BASETYPE`` flag. Without the basetype flag users cannot subclass an extension class and ``PyType_GetModule()`` will always return the module from ``PyType_FromModuleAndSpec``.

Reply:
> `_PyType_GetModuleByDef()` is only required when a C extension type has the `Py_TPFLAGS_BASETYPE` flag. Without the basetype flag users cannot subclass an extension class and `PyType_GetModule()` will always return the module from `PyType_FromModuleAndSpec`.

Got it; thanks